networks/rpc/endpoints.go

```diff
@@ -32,10 +32,10 @@ func StartHTTPEndpoint(endpoint string, apis []API, modules []string, cors []str
 		whitelist[module] = true
 	}
 	// Register all the APIs exposed by the services
-	handler := NewServer()
+	handler := NewServer(HTTPServer)
 	for _, api := range apis {
 		if whitelist[api.Namespace] || (len(whitelist) == 0 && api.Public) {
-			if err := handler.RegisterName(api.Namespace, api.Service); err != nil {
```

Comment:
I'm still not convinced about the `s.scheme != IPCServer` check in RegisterName.
Can't we simply add `api.Privileged` condition to the existing if statement?

```go
handler := NewServer()
for _, api := range apis {
    if !api.Privileged && ( whitelist[api.Namespace] || (len(whitelist) == 0 && api.Public) ) {
        if err := handler.RegisterName(api.Namespace, api.Service); err != nil {
            // ...
```

Reply:
For the `s.scheme != IPCServer`, please refer to the description of the PR. It is for the description. It is to allow the privileged RPC to only IPC. 

Now the condition for privileged value is located inside of the `RegisterName`. This is to delegate that mandatory check to the function, not the caller's scope. Please retake a look at the new changes. I hope to hear your feedback again about the new changes.

Reply:
If I understand correctly, it boils down to "if not InProc && api.Privileged, skip RegisterName". My conjecture is that the logic can be implemented by conditionally not calling `RegisterName`.

We can insert `if !api.Privileged` condition (1) only in StartHTTPEndpoint, StartFastHTTPEndpoint, StartWSEndpoint, StartFastWSEndpoint, startgRPC, (2) but not in StartIPCEndpoint, startInProc. Just like how `whitelist + api.Public` check is done now.

I think the `api.Public` check and `api.Privileged` check are very similar access controls - the names, purposes. Is there a reason why those two should be treated differently?

- Do you think the `Privileged` checks must be done inside `RegisterName`? 
- Do you think the `whitelist + Public` checks must be done inside `RegisterName`?
- Is there any other `RegisterName` call sites, where server scheme cannot be determined?
  - `MainBridge.SetComponents` seems to be the case. But then you set `mb.rpcServer = rpc.NewServer(rpc.InProcServer)`. Which means `MainBridge.SetComponents` is safe to ignore `api.Privileged` flag.

Reply:
I second blukat. I sense that you're worried about developers forgetting to check api.Privileged flag, but this seems like the tail is wagging the dog. A warning comment at `RegisterName` would suffice.

Reply:
Before putting my comment, let me correct your repeated incorrect name. 

- `InProc` is not `IPC`. `IPC` is to communicate via file scoket (`unix` or `window`).  `InProc` uses `process pipe` between two channels(pipe)`. 
-  Your condition is correct (`IPC && api.Privileged`).
- `Public` is imported from Ethereum, not me. The purpose of this property is described above in my comment. Thus, two properties are definitely different semantics.

Okay, now let me answer one by one:
> Do you think the Privileged checks must be done inside RegisterName?

The refactored code removes a few of lines that must appear before calling the `RegisterName()` in every call sites. Think logically, the privileged option was created to filter unprivileged RPCs. I think it should be handled by registration phase, which is a `RegisterName`. The validation in the caller seems not correct logically.

> Do you think the whitelist + Public checks must be done inside RegisterName? 

Sure, but I didn't touch the existing code for now.

> Is there any other RegisterName call sites, where server scheme cannot be determined?

The server scheme can be determined since the scheme parameter is passed to server initialization (`NewServer()`). The server scheme and privileged option belong to a different contexts. The scheme is checked inside of `RegisterName()`, and the privileged option is passed to `RegisterName()`. Please refer [1](https://github.com/klaytn/klaytn/pull/1369/files#diff-a0f8913808afa1e438d2664d0dbb0929a1ebc18873fb323b51f1f4a39e1d51c9R151) and [2](https://github.com/klaytn/klaytn/pull/1369/files#diff-a0f8913808afa1e438d2664d0dbb0929a1ebc18873fb323b51f1f4a39e1d51c9R242).

Reply:
> The refactored code removes a few of lines that must appear before calling the RegisterName() in every call sites. Think logically, the privileged option was created to filter unprivileged RPCs. I think it should be handled by registration phase, which is a RegisterName. The validation in the caller seems not correct logically.

If you want to prevent someone accidentally missing the `Privileged` check, then I understand why the checks should be inside `RegisterName`.

> Sure, but I didn't touch the existing code for now.

Okay, I agree that we'd better keep the existing code for now.

> The server scheme can be determined since the scheme parameter is passed to server initialization (NewServer()). The server scheme and privileged option belong to a different contexts. The scheme is checked inside of RegisterName(), and the privileged option is passed to RegisterName(). Please refer [1](https://github.com/klaytn/klaytn/pull/1369/files#diff-a0f8913808afa1e438d2664d0dbb0929a1ebc18873fb323b51f1f4a39e1d51c9R151) and [2](https://github.com/klaytn/klaytn/pull/1369/files#diff-a0f8913808afa1e438d2664d0dbb0929a1ebc18873fb323b51f1f4a39e1d51c9R242).

I wanted to say if the check is outside `RegisterName`, then we won't have to pass the `scheme` to `NewServer()`. But for the check to be inside `RegisterName`, of course I see that `scheme` should be passed. 

Personally I would prefer conciseness in this particular case. I think the risk of missing the check is small because there are only handful number of call sites and we can review the changes. To me, changes are too large compared to the feature.

However, being careful is also important. Therefore I am okay with your changes.

Reply:
Yes. Of course, the accidentally missing condition is also one of the concerns of my changes. However, the main purpose is to decide the logically correct place where the `priviliged` flag is looked up. By delegating to the function, the caller does not take care of any consideration. The function takes care of validation.